### ISSUE TITLE
Add a proper learning rate schedule to the BERT example

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-nlp#84]
We should add learning rate schedule matching the [original BERT implementation](https://github.com/google-research/bert/blob/master/optimization.py#L32) to our BERT example in `/examples/bert`.

**Comment by mattdangerw** [comment]
https://keras.io/api/callbacks/learning_rate_scheduler/

**Comment by Stealth-py** [comment]
Hey @mattdangerw! I'd like to work on this issue.

**Comment by Stealth-py** [comment]
Hey @mattdangerw! I just wanted to confirm a couple of things which I'll be doing:
- The file where the issue is focused is `/examples/bert/run_pretraining.py`.
- Adding a `scheduler` function as referenced [here](https://keras.io/api/callbacks/learning_rate_scheduler/) and then use the same to update the `learning_rate` like so: `learning_rate_schedule = scheduler(epoch = FLAGS.epochs, lr = FLAGS.learning_rate)`.
Am I correct?

**Comment by mattdangerw** [comment]
Thanks for contributing!

Yes this should be changing `/examples/bert/run_pretraining.py`

Re scheduler, sorry I linked the wrong resource there. We need to set the learning rate depending on the number of train steps, not epochs, so we should use the following:
https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/LearningRateSchedule

I believe we want a linear ramp up during the first 10k training steps, followed by a linear decay. The code I linked shows the logic for original BERT, though it will need to be adapted to work with Keras/TF2.

**Comment by Stealth-py** [comment]
Thank you for the review!

I think I understand, so currently I'm thinking of implementing the scheduler by increasing the learning rate with a factor of `initial_learning_rate/(current_step + 1)`, and for decaying, would a factor shallower than the increasing factor be fine, till an `end_learning rate` of let's say `0.0001`?

**Comment by Stealth-py** [comment]
Hey @mattdangerw, I made some changes post my previous comment. So, now the factors being used are `initial_learning_rate/1e3` to increase the learning rate till `10,000 steps` and `initial_learning_rate/1e4` to decrease the learning rate after reaching the threshold of `10,000 steps`.

**Comment by mattdangerw** [comment]
From my reading of that code, given an input `learning_rate` parameter... It looks like the learning rate should increase linearly from 0 to `learning_rate` over the first 10000 steps, and then decay linearly from `learning_rate` back to zero over the rest of training.

For the decay part, that would be tricky to do with the way we are currently specifying train time (in epochs). So we should probably switch to training for a fixed number of train steps (via `steps_per_epoch`) while we do this.

Overall something like:
 - Remove `epochs` script argument, release with `num_train_steps`.
 -  Add a `num_warmup_steps` argument.
 - Implement the ramp up and back down, using `tf.keras.optimizers.schedules.LearningRateSchedule`.

Does that make sense/map with your reading of the original BERT code?

**Comment by Stealth-py** [comment]
Hey, yeah I think I kind of get it. From what I understand, I have a few doubts and a few things I would like to confirm:
1. Should I remove the `epochs` flag itself, or should I just replace the argument from `model.fit` or will `model.fit_generator` be better as shown [here](https://github.com/keras-team/keras/issues/7506#issuecomment-322050382)? What values do I use for `num_train_steps` or `steps_per_epoch`, given that both of them are co-related to each other. Should I use a value similar to what was shown in the original BERT paper as shown in [this issue](https://github.com/google-research/bert/issues/1025).
2. So, the linear ramp-up/decay with the `num_warmup_steps` is fine and can be done using the `num_warmup_steps` and `num_train_steps`, which is not an issue for now.
3. HuggingFace implements it's optimizer as shown [here](https://huggingface.co/docs/transformers/main_classes/optimizer_schedules#transformers.get_linear_schedule_with_warmup), so I was thinking if this can be used as a reference.

**Comment by mattdangerw** [comment]
1. Yeah, good point. Calculating `steps_per_epoch` from the `epochs` flag and `num_train_steps` is probably a good way to do this.
3. Yeah, that looks like a great reference point.

**Comment by Stealth-py** [comment]
Alright, thanks for the review!
I'll wrap this up ASAP, and open a PR in a bit :D
